test/gtest/ucp/test_ucp_reconfigure.cc

```diff
@@ -0,0 +1,264 @@
+
+    tl_bitmap = is_exclude_iface ? UCS_STATIC_BITMAP_NOT(other.ep_tl_bitmap()) :
+                                   ucp_tl_bitmap_max;
+
+    UCS_ASYNC_BLOCK(&worker()->async);
+    ASSERT_UCS_OK(ucp_ep_create_to_worker_addr(worker(), &tl_bitmap,
+                                               &worker_addr->second,
+                                               UCP_EP_INIT_CREATE_AM_LANE,
+                                               "reconfigure test", addr_indices,
+                                               &m_ep));
+
+    m_ep->conn_sn = 0;
+    ASSERT_TRUE(ucp_ep_match_insert(worker(), m_ep, worker_addr->second.uuid,
+                                    m_ep->conn_sn, UCS_CONN_MATCH_QUEUE_EXP));
+
+    ASSERT_UCS_OK(ucp_wireup_send_request(m_ep));
+    UCS_ASYNC_UNBLOCK(&worker()->async);
+
+    store_config();
+}
+
+bool test_ucp_reconfigure::entity::has_matching_lane(ucp_ep_h ep,
+                                                     ucp_lane_index_t lane_idx,
+                                                     const entity &other) const
+{
+    const auto lane     = &ucp_ep_config(ep)->key.lanes[lane_idx];
+    auto context        = m_worker_entity->ucph();
+    const auto resource = &context->tl_rscs[ucp_ep_get_rsc_index(ep, lane_idx)];
+    auto addr           = other.get_address(true);
+    const ucp_address_entry_t *ae;
+
+    ucs_carray_for_each(ae, addr->second.address_list,
+                        addr->second.address_count) {
+        if ((resource->tl_name_csum == ae->tl_name_csum) &&
+            ucp_wireup_is_lane_connected(ep, lane_idx, ae)) {
+            EXPECT_EQ(ae->sys_dev, lane->dst_sys_dev);
+            EXPECT_EQ(ae->md_index, lane->dst_md_index);
+            return true;
+        }
+    }
+
+    return false;
+}
+
+void test_ucp_reconfigure::entity::verify_configuration(
+        const entity &other) const
+{
+    unsigned reused                  = 0;
+    const ucp_lane_index_t num_lanes = ucp_ep_num_lanes(ep());
+
+    EXPECT_EQ(num_lanes, ucp_ep_num_lanes(other.ep()));
+
+    for (ucp_lane_index_t lane = 0; lane < num_lanes; ++lane) {
+        /* Verify local and remote lanes are identical */
+        EXPECT_TRUE(has_matching_lane(ep(), lane, other));
+
+        /* Verify correct number of reused lanes is configured */
+        auto uct_ep = ucp_ep_get_lane(ep(), lane);
+        auto it     = std::find(m_uct_eps.begin(), m_uct_eps.end(), uct_ep);
+        reused     += (it != m_uct_eps.end());
+    }
+
+    EXPECT_EQ(reused, is_reconfigured() ? 0 : num_lanes);
+}
+
+test_ucp_reconfigure::address_p
+test_ucp_reconfigure::entity::get_address(bool ep_only) const
+{
+    const ucp_tl_bitmap_t tl_bitmap = ep_only ? ep_tl_bitmap() :
+                                                ucp_tl_bitmap_max;
+    size_t addr_len;
+
+    address_p address(new address_t(nullptr, {0}), [](address_t *addr) {
+        ucs_free(addr->first);
+        ucs_free(addr->second.address_list);
+        delete addr;
+    });

```

Comment:
it looks unreadable, maybe add a wrapper class with clear API?

Reply:
why unreachable? IMO wrapper class is overhead here

Reply:
ok, up to other reviewers, it looks ugly to me

Reply:
As a compromise, we can keep this code but add some doc above.

Reply:
Unresolved

Reply:
line 219